pkg/resource/deploy/step_generator.go

```diff
@@ -2154,6 +2154,17 @@ func (sg *stepGenerator) calculateDependentReplacements(root *resource.State) ([
 			NewInputs:     inputsForDiff,
 			AllowUnknowns: true,
 		})
+
+		goal, hasGoal := sg.deployment.goals.Load(r.URN)
```

Comment:
Well this doesn't work as expected because the goal for the `RegisterResourceEvent` for the dependent resource can come in after the one for the resource that we're calculating dependent replaces for.

Back to the drawing board

Reply:
I have the suspicion that this will need https://github.com/pulumi/pulumi/issues/7285